File: leetcode/p00094/p00094.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        if not root:
            return res
        
        visited = set()
        stack = [root]
        while len(stack) > 0:
            node = stack.pop()

            if node.left and node.left not in visited:
                stack.append(node)
                stack.append(node.left)
            else:
                res.append(node.val)
                visited.add(node)
                if node.right and node.right not in visited:
                    stack.append(node.right)
        
        return res

File: leetcode/p00094/p00094.go
package p00094

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func inorderTraversal(root *TreeNode) []int {
	res := []int{}
	if root == nil {
		return res
	}

	s := CreateStack()
	for root != nil || !s.IsEmpty() {
		for root != nil {
			s.Put(root)
			root = root.Left
		}
		root = s.Pop()
		res = append(res, root.Val)
		root = root.Right
	}

	return res
}

type Stack struct {
	arr  []*TreeNode
	size int
}

func CreateStack() Stack {
	return Stack{
		arr:  make([]*TreeNode, 0),
		size: 0,
	}
}

func (s *Stack) IsEmpty() bool {
	return s.size == 0
}

func (s *Stack) Put(node *TreeNode) {
	s.arr = append(s.arr, node)
	s.size++
}

func (s *Stack) Pop() *TreeNode {
	if s.IsEmpty() {
		return nil
	}

	node := s.arr[s.size-1]
	s.arr = s.arr[:s.size-1]
	s.size--

	return node
}
